File: 472.concatenated-words.java
/*
 * @lc app=leetcode id=472 lang=java
 *
 * [472] Concatenated Words
 */

// @lc code=start
class Solution {
    // use trie to check concatenated words
    Trie trie = new Trie();

    public List<String> findAllConcatenatedWordsInADict(String[] words) {
        List<String> res = new ArrayList<String>();
        // 1. for convience, sort it length-ascendingly
        Arrays.sort(words, (a, b) -> a.length() - b.length());
        // 2. walk thourgh all the words, recursively check if it is concatenated
        for (int i = 0; i < words.length; i++) {
            String word = words[i];
            boolean[] visited = new boolean[word.length()];
            if (isConcatenated(word, 0, visited)) {
                res.add(word);
            } else {
                insert2Trie(word);
            }
        }
        return res;
    }

    public boolean isConcatenated(String word, int start, boolean[] visited) {
        if (word.length() == start) {
            return true;
        }
        if (visited[start]) {
            return false;
        }

        visited[start] = true;
        Trie node = trie;
        for (int i = start; i < word.length(); i++) {
            char currChar = word.charAt(i);
            node = node.children[currChar - 'a'];

            if (node == null) {
                return false;
            }
            if (node.isEnd) {
                if (isConcatenated(word, i + 1, visited)) {
                    return true;
                }
            }
        }
        return false;
    }

    // insert word to Trie
    public void insert2Trie(String word) {
        Trie node = trie;
        for (int i = 0; i < word.length(); i++) {
            char currChar = word.charAt(i);

            if (node.children[currChar - 'a'] == null) {
                node.children[currChar - 'a'] = new Trie();
            }
            node = node.children[currChar - 'a'];
        }
        node.isEnd = true;
    }
}

class Trie {
    Trie[] children;
    boolean isEnd;

    public Trie() {
        children = new Trie[26];
        isEnd = false;
    }
}

// @lc code=end

File: 472.concatenated-words.py
#
# @lc app=leetcode id=472 lang=python3
#
# [472] Concatenated Words
#

# @lc code=start


class Solution:
    def findAllConcatenatedWordsInADict(self, words: List[str]) -> List[str]:
        word_set = set(words)
        result = []

        for word in words:
            if not word:
                continue
            # Remove the word itself to prevent using it in its own construction
            word_set.remove(word)
            if self.canForm(word, word_set):
                result.append(word)
            # Add it back for the next iteration
            word_set.add(word)

        return result

    def canForm(self, word, word_set):
        n = len(word)
        # dp[i] means word[:i] can be formed by words in word_set
        dp = [False] * (n + 1)
        dp[0] = True

        for i in range(n):
            if not dp[i]:
                continue
            # Try every possible next cut
            for j in range(i + 1, n + 1):
                # If the substring word[i:j] exists in the set,
                # mark dp[j] = True
                if word[i:j] in word_set:
                    dp[j] = True
                    # Early exit: if we've reached the end, it's formable
                    if j == n:
                        return True
        return False


# @lc code=end
